### DISCUSSION TITLE
Current U-boot configuration is not friendly for smaller flash

### DISCUSSION THREAD

**Discussion by ltrsunny** [linux-chenxing/linux-chenxing.org#55] [General]
Like SOM2D01, there are resistors to set NAND or NOR flash, but all the related options are force selected UBI to store u-boot and environment. (in *dts, a 128MB MTD partition is defined which could be larger than many flash)

For better compatibility, I think some of the options (like `ENV_IS_IN_UBI` & `SPL_UBI`) shouldn't be force enabled.

**Comment by fifteenhex** [comment]
I thought the SOM2D01 would always have NAND... what if I add an option to select if NOR or NAND is mounted and select the right options?

I prefer to it KConfig instead of defconfigs as the defconfigs are pain to update

**Reply by ltrsunny** [reply]
> I thought the SOM2D01 would always have NAND... what if I add an option to select if NOR or NAND is mounted and select the right options?
> 
> I prefer to it KConfig instead of defconfigs as the defconfigs are pain to update

Actually they have the optional design in hardware.
![image](https://user-images.githubusercontent.com/49466661/149612302-978f128d-eadb-448d-98dd-68524cf951f5.png)
and I have confirmed with the vendor that they can provide customized flash for batch order.

I agree with you, KConfig is a better method, but UBI related options shouldn't be force for this platform.

**Reply by fifteenhex** [reply]
> Actually they have the optional design in hardware and I have confirmed with the vendor that they can provide customized flash for batch order.

That's good to know... we might need to consider that in the kernel side too.

> I agree with you, KConfig is a better method, but UBI related options shouldn't be force for this platform.

I can add a menu in the configuration where you can select the type of flash and that will select if UBI is enabled or not.

**Reply by ltrsunny** [reply]
> > Actually they have the optional design in hardware and I have confirmed with the vendor that they can provide customized flash for batch order.
> 
> That's good to know... we might need to consider that in the kernel side too.
> 
> > I agree with you, KConfig is a better method, but UBI related options shouldn't be force for this platform.
> 
> I can add a menu in the configuration where you can select the type of flash and that will select if UBI is enabled or not.

Great

**Reply by ltrsunny** [reply]
> 
> I can add a menu in the configuration where you can select the type of flash and that will select if UBI is enabled or not.

@fifteenhex Daniel I want to confirm something, does NAND only have the UBI method to load u-boot in current SPL configurations?
I am trying to make a small image for production which only contains bootloader part. I tried `SPL_SPI_LOAD` but it seems only for NOR.
If I use the UBI to store the u-boot, I cannot find where is the exact address for the UBI partition by nanddump.

**Reply by fifteenhex** [reply]
NAND should use UBI for every because NAND can have bad blocks and needs ECC etc.
You can create an image with GCIS, IPL, u-boot SPL and the ubi partition containing u-boot. But it takes a long time to write via i2c.

**Reply by ltrsunny** [reply]
> NAND should use UBI for every because NAND can have bad blocks and needs ECC etc.

Got it. New knowledge to me.
> But it takes a long time to write via i2c.

I think I can solve this by modifying the dts to get a small mtd1 and create UBI over it.
Another solution is to use mmc in SPL to load uboot from SD（or use Falcon mode to call the kernel directly). I don't know it can be a image file or must be a partition. I am trying on it.

Do you have nicer idea?

**Reply by ltrsunny** [reply]
Actually in official SDK, they wrote the uboot image(200k+) directly into nand and ubi is only for linux.

Current uboot compiled image is around 400K, will it be dangerous to write such a small part?
